reinforced_lib/rlib.py

```diff
@@ -0,0 +1,337 @@
+        -------
+        out : gym.spaces.Space
+            Action space of selected agent.
+        """
+
+        if not self._agent:
+            raise NoAgentError()
+
+        return self._agent.action_space()
+
+    def init(self, seed: int = None) -> int:
+        """
+        Initializes new instance of the agent.
+
+        Parameters
+        ----------
+        seed : int
+            A number used to initialize the JAX pseudo-random number generator.
+
+        Returns
+        -------
+        out : int
+            The identifier of created instance.
+        """
+
+        agent_id = len(self._agents_states)
+        seed = seed if seed else 0
+
+        self._agents_states.append(self._agent.init())
+        self._agents_keys.append(jax.random.PRNGKey(seed))
+
+        return agent_id
+
+    def sample(
+            self,
+            agent_id: int = 0,
+            *args,
+            update_observations: Union[Dict, Tuple, Any] = None,
+            sample_observations: Union[Dict, Tuple, Any] = None,
+            **kwargs
+    ) -> Any:
+        """
+        Takes the environment state as input, updates the agent state, and returns the next action selected by 
+        the agent. If 'no_env_mode' is disabled, observations are passed by *args and **kwargs (observations must
+        match selected environment observation space). If 'no_env_mode' is enabled, observations must be passed 
+        by 'update_observations' and 'sample_observations' parameters (observations must match agents 
+        'update_observation_space' and 'sample_observation_space'). If there are no agent instance initialized,
+        the method automatically initializes the first instance.
+
+        Parameters
+        ----------
+        agent_id : int
+            The identifier of agent instance.
+        args : Tuple
+            Environment observations.
+        update_observations : Union[Dict, Tuple, Any]
+            Observations used when 'no_env_mode' is enabled (must match agents 'update_observation_space').
+        sample_observations : Union[Dict, Tuple, Any]
+            Observations used when 'no_env_mode' is enabled (must match agents 'sample_observation_space').
+        kwargs : Dict
+            Environment observations.
+
+        Returns
+        -------
+        out : Any
+            Action selected by the agent.
+        """
+
+        if not self._agent:
+            raise NoAgentError()
+
+        if not self._no_env_mode and not self._env:
+            raise NoEnvironmentError()
+
+        if len(self._agents_states) == 0:
+            self.init()
+
+        self._agents_keys[agent_id], key = jax.random.split(self._agents_keys[agent_id])
+        state = self._agents_states[agent_id]
+
+        if not self._no_env_mode:
+            update_observations, sample_observations = self._env.transform(*args, **kwargs)
+
+        if isinstance(update_observations, dict):
+            state = self._agent(state, **update_observations)
```

Comment:
czy nie zgubiłeś tutaj metody `update`? Tzn czy nie powinno być `state = self._agent.update(state, **update_observations)`, i analogicznie w `elif` i `else`?

Reply:
Tak jest, zorientowałem się dopiero przy implementacji pull requesta #6 i tam to poprawiłem.

Reply:
super